scapy/packet.py

```diff
@@ -1694,39 +1696,90 @@ def decode_payload_as(self, cls):
-        obtain the same packet
+        Internal method used to generate command() and toJSON()
         """
         f = []
-        for fn, fv in self.fields.items():
+        iterator: Iterator[Tuple[str, Any]]
+        if json:
+            iterator = ((x.name, self.getfieldval(x.name)) for x in self.fields_desc)
+        else:
+            iterator = iter(self.fields.items())
+        for fn, fv in iterator:
             fld = self.get_field(fn)
             if isinstance(fv, (list, dict, set)) and not fv and not fld.default:
                 continue
             if isinstance(fv, Packet):
-                fv = fv.command()
+                if json:
+                    fv = {k: v for (k, v) in fv._command(json=True)}
+                else:
+                    fv = fv.command()
             elif fld.islist and fld.holds_packets and isinstance(fv, list):
-                fv = "[%s]" % ",".join(map(Packet.command, fv))
+                if json:
+                    fv = [
+                        {k: v for (k, v) in x}
+                        for x in map(lambda y: Packet._command(y, json=True), fv)
+                    ]
+                else:
+                    fv = "[%s]" % ",".join(map(Packet.command, fv))
             elif fld.islist and isinstance(fv, list):
-                fv = "[%s]" % ", ".join(
-                    getattr(x, 'command', lambda: repr(x))()
-                    for x in fv
-                )
+                if json:
+                    fv = [
+                        getattr(x, 'command', lambda: repr(x))()
+                        for x in fv
+                    ]
+                else:
+                    fv = "[%s]" % ",".join(
+                        getattr(x, 'command', lambda: repr(x))()
+                        for x in fv
+                    )
             elif isinstance(fv, FlagValue):
                 fv = int(fv)
             elif callable(getattr(fv, 'command', None)):
-                fv = fv.command()
+                fv = fv.command(json=json)
             else:
-                fv = repr(fld.i2h(self, fv))
-            f.append("%s=%s" % (fn, fv))
-        c = "%s(%s)" % (self.__class__.__name__, ", ".join(f))
+                if json:
+                    if isinstance(fv, bytes):
+                        fv = fv.decode("utf-8", errors="backslashreplace")
+                    else:
+                        fv = fld.i2h(self, fv)
+                else:
+                    fv = repr(fld.i2h(self, fv))
+            f.append((fn, fv))
+        return f
+
+    def command(self):
+        # type: () -> str
+        """
+        Returns a string representing the command you have to type to
+        obtain the same packet
+        """
+        c = "%s(%s)" % (
+            self.__class__.__name__,
+            ", ".join("%s=%s" % x for x in self._command())
+        )
         pc = self.payload.command()
         if pc:
             c += "/" + pc
         return c
 
+    def toJSON(self):
```

Comment:
Small nit: could ˋjson()‘ be a simpler name?

Reply:
Yeah you're right